### PR TITLE
[bugfix] Add end timestamp to TimeWindowPartitionsDefinition repr

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Adds `end` value to the repr of `TimeWindowPartitionsDefinition`. This fixes an issue where we could have non-deterministic assignment of assets to base asset jobs, leading to execution errors, if two different `TimeWindowPartitionsDefinition` with the same start date but different end dates were present.

Possible fix for #21152.

## How I Tested These Changes

New unit test for partitions def sorting.

### CODE REVIEW

**Comment by smackesey** [comment]
* **#21497** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21497?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by schrockn** [comment]
Can you find and note the blame PR where we added the sort key based on `repr` and link to it from here?

File: python_modules/dagster/dagster/_core/definitions/time_window_partitions.py [hunk 1]
```diff
@@ -488,6 +488,7 @@ def __repr__(self):
         # Replaces start time with timestamp as a workaround to make sure the repr is consistent across versions.
         return (
             f"TimeWindowPartitionsDefinition(start={self.start.timestamp()},"
+            f" end={self.end.timestamp() if self.end else None},"
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/definitions/time_window_partitions.py:491]: if we aren't going to be able to do a fix where we don't use `__repr__` for sorting soon, can you add a note to the attrs for `TimeWindowPartitionsDefinition` saying to update `__repr__` for any added attr?

**Review by jamiedemaria — APPROVED** [review]

**Comment by schrockn** [comment]
I have been inspired to start a catalog of anti-patterns and code smells: https://github.com/dagster-io/internal/discussions/9505

**Comment by alangenfeld** [comment]
> Can you find and note the blame PR where we added the sort key based on repr and link to it from here?

I went digging out of curiosity, believe this is the origin https://github.com/dagster-io/dagster/commit/3b55c4e864775b7a70ed8ff539629317a1202505 then it moved around a few times

**Comment by schrockn** [comment]
Thanks Alex. PR is https://github.com/dagster-io/dagster/pull/7388

**Pull request merged**